File: src/Test/LazyEvaluationFFI.fs
module Test.LazyEvaluationFFI

type Lazy<'a> = Lazy of (unit -> 'a)
let defer thunk = Lazy thunk
let force (Lazy thunk) = thunk ()
let rec buildThunks depth initial =
    if depth = 0 then initial
    else buildThunks (depth - 1) (defer (fun () -> force initial + 1))
let runLazyEvaluationFFI (input: obj) =
    let rec go count acc =
        if count = 0 then acc
        else go (count - 1) (acc + force (buildThunks 1000 (defer (fun () -> 0))))
    go (unbox<int> input) 0 :> obj

File: src/Test/LazyEvaluationFFI.java
    @FunctionalInterface
    private interface Lazy<A> { A force(); }

    private static <A> Lazy<A> defer(java.util.function.Supplier<A> thunk) {
        return () -> thunk.get();
    }

    private static <A> A force(Lazy<A> value) { return value.force(); }

    private static Lazy<Integer> buildThunks(int depth, Lazy<Integer> initial) {
        Lazy<Integer> result = initial;
        for (int i = 0; i < depth; i++) {
            Lazy<Integer> previous = result;
            result = defer(() -> force(previous) + 1);
        }
        return result;
    }

    public static final java.util.function.Function<Object, Object> runLazyEvaluationFFI = Bench.nativeBenchmark(input -> {
        int count = (Integer) input;
        int result = 0;
        for (int i = 0; i < count; i++) {
            result += force(buildThunks(1000, defer(() -> 0)));
        }
        return result;
    });
